### PR TITLE
feat: Job deadlines

### LINKED ISSUE #80: Feat: A mechanism to set activeDeadlineSeconds
We don't want to allow tracejobs to accumulate without bound or accidentally left running, sapping system resources for traces no one will look at.

To achieve this, I think that setting `activeDeadlineSeconds` on the trace job would provide a built-in way to ensure that no job runs longer than a sane default (say, 1 hour), unless specifically requested with a command-line flag to override the deadline (eg, --deadline 24h) to extend the time that a job is allowed to run.

Ideally, this would also signal bpftrace to dump its map / summarize the map data before exiting, and could also be used as a mechanism for having jobs run for a specific duration. Eg, sample for 10 minutes than exit.

### PR DESCRIPTION
**Pull request by dalehamel**
Fixes #80 and paves the way for #11 

I think this is pretty elegant and I have to give the credit for both ideas to @jerr 

This allows for ensuring that bpftrace is signalled with SIGINT to dump its map before exiting.

Have tested this and it works more reliably than the interactive traces via TTY attach.

A nice benefit is you can now collect data for a pre-set interval before exiting 😂

### CODE REVIEW

File: pkg/tracejob/job.go [hunk 1]
```diff
@@ -294,6 +296,21 @@ func (t *TraceJobClient) CreateJob(nj TraceJob) (*batchv1.Job, error) {
...
+							// we will also wait for an arbitrary amount of time (10s) to give bpftrace time to
+							// process and summarize the data
+							// FIXME should this sleep be configurable? 10s is probably ample for most cases.
+							Lifecycle: &apiv1.Lifecycle{
+								PreStop: &apiv1.Handler{
+									Exec: &apiv1.ExecAction{
+										Command: []string{
+											"/bin/bash",
+											"-c",
+											"kill -SIGINT $(pidof bpftrace) && sleep 10",
```
Comment (dalehamel) [inline pkg/tracejob/job.go]: Probably shouldn't embed this magic number in the string. Should I go so far as to make this configurable? (deadline-grace-period?)

Comment (dalehamel) [inline pkg/tracejob/job.go]: Probably shouldn't embed this magic number in the string. Should I go so far as to make this configurable? (deadline-grace-period?)

File: pkg/tracejob/job.go [hunk 2]
```diff
@@ -294,6 +297,21 @@ func (t *TraceJobClient) CreateJob(nj TraceJob) (*batchv1.Job, error) {
 							SecurityContext: &apiv1.SecurityContext{
 								Privileged: boolPtr(true),
 							},
+							// We want to send SIGINT prior to the pod being killed, so we can print the map
+							// we will also wait for an arbitrary amount of time (10s) to give bpftrace time to
+							// process and summarize the data
+							// FIXME should this sleep be configurable? 10s is probably ample for most cases.
```
Comment (leodido) [inline pkg/tracejob/job.go]:
```suggestion
```

This comment line can be removed now, imho!

**Comment by dalehamel** [comment]
thanks @leodido 

BTW one side-effect of this is that anything that trips the job deadline will appear as a failed job. That's not ideal, but if we check the exit status of the container (ie, if it actually succeeded) before the pod is GC'd we should be able to rescue the true exit status.

For now i don't think it much matters if a job passed its deadline reports as failed, even if this was expected. One way around this would be to have the tracejob runner time out, so that the job exits cleaning and shows as completed if this happens.

I think that can be addressed in a separate PR though unless anyone feels strongly.

**Comment by dalehamel** [comment]
Before merging I want to try and let the trace runner die gracefully, I don't feel right making 'failed jobs' a norm

File: pkg/tracejob/job.go [hunk 3]
```diff
@@ -184,6 +187,11 @@ func (t *TraceJobClient) DeleteJobs(nf TraceJobFilter) error {
 func (t *TraceJobClient) CreateJob(nj TraceJob) (*batchv1.Job, error) {
 
 	bpfTraceCmd := []string{
+		"/bin/timeout",
```
Comment (dalehamel) [inline pkg/tracejob/job.go]: I'm pretty happy with this solution, I tested it out and it allows trace jobs to complete within their deadline and print their maps. It's also a more reliable way to ensure maps actually do get printed, and access their log data.

**Comment by dalehamel** [comment]
@leodido can you take another look please?

I basically just added the timeout command, and upped the k8s deadline to include the grace period. This should give plenty of time for the process to shut down cleanly, so it doesn't need to rely on the pre-stop hook.

So, if a job times out, we will have a failed job that is past its activity deadline, similar to exit 124 from the timeout command, indicating a job did actually pass its deadline and didn't exit as it was supposed to. This should ideally be a rare case.

In most cases, we should see that the job is able to complete and get the output from the logs, even if it is a map or histogram.

**Pull request merged**